File: python-leetcode/backtracking/131.分割回文串.py
from typing import List


class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        path = []

        def backtrack(startIndex: int):
            if startIndex == len(s):
                res.append(path[:])
                return
            for i in range(startIndex, len(s)):
                # startIndex - i 是否为回文
                if not self.isPalindrome(s, startIndex, i):
                    continue
                subStr = s[startIndex : i + 1]
                path.append(subStr)
                backtrack(i + 1)
                # 回溯
                path.pop()

        backtrack(0)
        return res

    def isPalindrome(self, s: str, l: int, r: int) -> bool:
        while l < r:
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        return True

File: go-leetcode/backtracking/131.分割回文串.go
package leetcode

func partition(s string) [][]string {
	res := [][]string{}
	path := []string{}

	var backtrack func(startIndex int)
	backtrack = func(startIndex int) {
		if startIndex == len(s) {
			tmp := make([]string, len(path))
			copy(tmp, path)
			res = append(res, tmp)
			return
		}
		for i := startIndex; i < len(s); i++ {
			// startIndex - i 是否为回文
			if !isPalindrome(s, startIndex, i) {
				continue
			}
			subStr := s[startIndex : i+1]
			path = append(path, subStr)
			backtrack(i + 1)
			// 回溯
			path = path[:len(path)-1]
		}
	}

	backtrack(0)
	return res
}

func isPalindrome(s string, l, r int) bool {
	for l < r {
		if s[l] != s[r] {
			return false
		}
		l++
		r--
	}
	return true
}
